agent/consul/leader_metrics.go

```diff
@@ -43,6 +58,83 @@ func rootCAExpiryMonitor(s *Server) certExpirationMonitor {
 	}
 }
 
+func primaryCAExpiryMonitor(s *Server) certExpirationMonitor {
+	return certExpirationMonitor{
+		Key: metricsKeyMeshPrimaryCAExpiry,
+		Labels: []metrics.Label{
+			{Name: "datacenter", Value: s.config.Datacenter},
+		},
+		Logger: s.logger.Named(logging.Connect),
+		Query: func() (time.Duration, error) {
+
+			isPrimary := s.config.Datacenter == s.config.PrimaryDatacenter
+			if isPrimary {
+				provider, _ := s.caManager.getCAProvider()
+
+				if _, ok := provider.(ca.PrimaryUsesIntermediate); !ok {
+					cert, err := getActiveIntermediate(s)
+					if err != nil {
+						return 0, err
+					}
+					return time.Until(cert.NotAfter), nil
+				}
+
+				state := s.fsm.State()
+				_, root, err := state.CARootActive(nil)
+				switch {
+				case err != nil:
+					return 0, fmt.Errorf("failed to retrieve root CA: %w", err)
+				case root == nil:
+					return 0, fmt.Errorf("no active root CA")
+				}
+
+				return time.Until(root.NotAfter), nil
+			}
+			return 0, nil
+		},
+	}
+}
+
+func secondaryCAExpiryMonitor(s *Server) certExpirationMonitor {
+	return certExpirationMonitor{
+		Key: metricsKeyMeshSecondaryCAExpiry,
+		Labels: []metrics.Label{
+			{Name: "datacenter", Value: s.config.Datacenter},
+		},
+		Logger: s.logger.Named(logging.Connect),
+		Query: func() (time.Duration, error) {
+			isPrimary := s.config.Datacenter == s.config.PrimaryDatacenter
+			if !isPrimary {
+				cert, err := getActiveIntermediate(s)
+				if err != nil {
+					return 0, err
+				}
+				return time.Until(cert.NotAfter), nil
+			}
+			return 0, nil
+		},
+	}
+}
+
+func getActiveIntermediate(s *Server) (*x509.Certificate, error) {
+	state := s.fsm.State()
+	_, root, err := state.CARootActive(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// the CA used in a secondary DC is the active intermediate,
+	// which is the last in the IntermediateCerts stack
+	if len(root.IntermediateCerts) == 0 {
+		return nil, errors.New("no intermediate available")
+	}
+	cert, err := connect.ParseCert(root.IntermediateCerts[len(root.IntermediateCerts)-1])
```

Comment:
I'm not familiar enough with this code, but I wonder how safe this assumption is (that the last cert will be the important one). This may be fine just calling out as something I need to look into more to confirm.

Reply:
I went through all the code that generate an intermediate and modify this and it always append to it. So the most recent cert will be always the last and this is the one we want to expose.